core/src/main/scala/filodb.core/memstore/PartKeyLuceneIndex.scala

```diff
@@ -289,18 +298,58 @@ class PartKeyLuceneIndex(ref: DatasetRef,
     indexWriter.close()
   }
 
-  // TODO - most frequently used eviction needed
-  private val readerStateCache = mutable.WeakHashMap[(IndexReader, String), DefaultSortedSetDocValuesReaderState]()
+  // The reader state object is expensive to create, so reuse across queries.
+  // It is associated with reader/labelName
+  private val readerStateCache: LoadingCache[(IndexReader, String), DefaultSortedSetDocValuesReaderState] =
+    // TODO cache config is currently hardcoded - make configuration properties if really deemed necessary
+    Caffeine.newBuilder()
+      .maximumSize(300)
+      .recordStats()
+      .expireAfter(new Expiry[(IndexReader, String), DefaultSortedSetDocValuesReaderState]() {
+        // shard key columns have larger expiry time
+        override def expireAfterCreate(key: (IndexReader, String),
+                                       value: DefaultSortedSetDocValuesReaderState,
+                                       currentTime: Long): Long = {
+          if (schema.options.shardKeyColumns.contains(key._2)) {
+            TimeUnit.HOURS.convert(5, TimeUnit.NANOSECONDS)
+          } else {
+            TimeUnit.MINUTES.convert(10, TimeUnit.NANOSECONDS)
+          }
+        }
+
+        override def expireAfterUpdate(key: (IndexReader, String),
+                                       value: DefaultSortedSetDocValuesReaderState,
+                                       currentTime: Long, currentDuration: Long): Long = {
+          if (schema.options.shardKeyColumns.contains(key._2)) {
+            TimeUnit.HOURS.convert(5, TimeUnit.NANOSECONDS)
+          } else {
+            TimeUnit.MINUTES.convert(10, TimeUnit.NANOSECONDS)
+          }
+        }
+
+        override def expireAfterRead(key: (IndexReader, String),
+                                     value: DefaultSortedSetDocValuesReaderState,
+                                     currentTime: Long, currentDuration: Long): Long = {
+          if (schema.options.shardKeyColumns.contains(key._2)) {
+            TimeUnit.HOURS.convert(5, TimeUnit.NANOSECONDS)
+          } else {
+            TimeUnit.MINUTES.convert(10, TimeUnit.NANOSECONDS)
+          }
+        }
+      })
+      .build((key: (IndexReader, String)) => {
+        new DefaultSortedSetDocValuesReaderState(key._1, FACET_FIELD_PREFIX + key._2)
+      })
+
   def labelValuesEfficient(colFilters: Seq[ColumnFilter], startTime: Long, endTime: Long,
                            colName: String, limit: Int = 100): Seq[String] = {
     require(facetEnabled, "Faceting not enabled on this index; labelValuesEfficient cannot be called")
     val labelValues = mutable.ArrayBuffer[String]()
+    val start = System.currentTimeMillis()
```

Comment:
using ``nanoTime`` for latency measurement is reliable

Reply:
done